File: App源码/ai_stock_picker/lib/services/news_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// 东方财富新闻服务 - 实时财经资讯
class NewsService {
  static const String _baseUrl = 'https://np-listapi.eastmoney.com/comm/web/getListInfo';
  static final http.Client _client = http.Client();

  /// 获取最新财经新闻
  /// [pageSize] 每页条数，默认10条
  /// [type] 1=财经要闻
  static Future<List<Map<String, dynamic>>> fetchLatestNews({int pageSize = 10}) async {
    try {
      final resp = await _client.get(
        Uri.parse('$_baseUrl?client=web&pageSize=$pageSize&type=1&mTypeAndCode=1.000001'),
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
          'Referer': 'https://so.eastmoney.com/',
        },
      ).timeout(const Duration(seconds: 15));

      if (resp.statusCode != 200 || resp.body.isEmpty) return [];

      final data = json.decode(resp.body);
      if (data['code'] != 1) return [];

      final list = data['data']?['list'] as List? ?? [];
      return list.map<Map<String, dynamic>>((item) {
        return {
          'title': item['Art_Title']?.toString() ?? '',
          'code': item['Art_Code']?.toString() ?? '',
          'time': item['Art_ShowTime']?.toString() ?? '',
          'url': item['Art_Url']?.toString() ?? '',
        };
      }).where((item) => item['title']!.isNotEmpty).toList();
    } catch (e) {
      return [];
    }
  }

  /// 抓取新闻正文内容
  /// 从东方财富新闻页面提取正文
  static Future<String> fetchNewsContent(String url) async {
    try {
      final resp = await _client.get(
        Uri.parse(url),
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
          'Referer': 'https://so.eastmoney.com/',
        },
      ).timeout(const Duration(seconds: 15));

      if (resp.statusCode != 200 || resp.body.isEmpty) {
        return '无法加载新闻内容，请检查网络连接。';
      }

      final html = resp.body;

      // 匹配 id="ContentBody" 的正文div
      final contentMatch = RegExp(
        "<div[^>]*id=[\"']ContentBody[\"'][^>]*>(.*?)</div>",
        caseSensitive: false,
        dotAll: true,
      ).firstMatch(html);

      if (contentMatch == null) {
        return '暂无新闻正文内容。';
      }

      // 去除HTML标签
      String content = contentMatch.group(1) ?? '';
      content = content.replaceAll(RegExp(r'<[^>]+>'), '');
      // 去除多余空白
      content = content.replaceAll(RegExp(r'\s+'), ' ').trim();
      // 去除文章来源
      content = content.replaceAll(RegExp(r'（文章来源：[^）]+）'), '');

      return content.isNotEmpty ? content : '暂无新闻正文内容。';
    } catch (e) {
      return '加载新闻内容失败：$e';
    }
  }

  /// 格式化时间显示
  static String formatTime(String timeStr) {
    try {
      final dt = DateTime.parse(timeStr);
      final now = DateTime.now();
      final diff = now.difference(dt);

      if (diff.inMinutes < 1) return '刚刚';
      if (diff.inMinutes < 60) return '${diff.inMinutes}分钟前';
      if (diff.inHours < 24) return '${diff.inHours}小时前';
      return '${dt.month}月${dt.day}日 ${dt.hour.toString().padLeft(2, '0')}:${dt.minute.toString().padLeft(2, '0')}';
    } catch (_) {
      return timeStr;
    }
  }
}

File: App源码/ai_stock_picker/backend/ai_stock_picker_backend/services/news_service.py
"""
东方财富新闻服务 - 热点追踪新闻源
"""
import httpx


class NewsService:
    """东方财富实时财经新闻"""

    def __init__(self):
        self._client = httpx.AsyncClient(timeout=15.0)
        self._base_url = "https://np-listapi.eastmoney.com/comm/web/getListInfo"

    async def close(self):
        await self._client.aclose()

    async def fetch_latest_news(self, page_size: int = 10) -> list[dict]:
        """获取最新财经新闻"""
        try:
            resp = await self._client.get(
                f"{self._base_url}?client=web&pageSize={page_size}&type=1&mTypeAndCode=1.000001",
                headers={
                    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
                    "Referer": "https://so.eastmoney.com/",
                },
            )
            if resp.status_code != 200 or not resp.text:
                return []
            data = resp.json()
            if data.get("code") != 1:
                return []
            items = data.get("data", {}).get("list") or []
            return [
                {
                    "title": str(item.get("Art_Title", "")),
                    "code": str(item.get("Art_Code", "")),
                    "time": str(item.get("Art_ShowTime", "")),
                    "url": str(item.get("Art_Url", "")),
                }
                for item in items
                if str(item.get("Art_Title", ""))
            ]
        except Exception:
            return []

    def filter_hot_news(self, news_list: list[dict]) -> list[dict]:
        """关键词初筛高潜力新闻"""
        hot_keywords = [
            # 政策类
            r'重大政策', r'政策转向', r'国务院', r'发改委', r'工信部', r'商务部',
            r'扶持', r'补贴', r'减免', r'放宽', r'准入',
            # 突发事件
            r'突发', r'紧急', r'限制出口', r'禁止出口', r'出口管制',
            r'制裁', r'断供', r'禁令', r'反制',
            # 从0到1新技术
            r'突破', r'首创', r'全球首', r'国内首', r'量产',
            r'商业化', r'落地', r'重大进展', r'关键突破',
            # 供需危机
            r'涨价', r'紧缺', r'缺货', r'断供', r'产能不足',
            r'供给收缩', r'需求暴增', r'供不应求',
            # 行业重大事件
            r'重组', r'并购', r'注入', r'借壳',
            r'中标', r'签约', r'大单', r'订单',
            # 市场异动
            r'涨停', r'暴涨', r'大爆发', r'掀涨停潮',
        ]

        filtered = []
        import re
        pattern = re.compile("|".join(hot_keywords))
        for news in news_list:
            title = news.get("title", "")
            if pattern.search(title):
                filtered.append(news)

        # 如果筛选后为空，返回前5条作为降级
        if not filtered:
            return news_list[:5]

        return filtered
